mls-rs-crypto-awslc/src/kem/ml_kem.rs

```diff
@@ -0,0 +1,235 @@
+                enc.as_ptr() as *mut u8,
+                enc.len(),
+            );
+
+            EVP_PKEY_CTX_free(ctx);
+            EVP_PKEY_free(pkey);
+
+            res
+        };
+
+        (res == 1)
+            .then_some(shared_secret)
+            .ok_or(Unspecified.into())
+    }
+
+    fn generate_deterministic(
+        &self,
+        ikm: &[u8],
+    ) -> Result<(HpkeSecretKey, HpkePublicKey), Self::Error> {
+        let nid = self.ml_kem.algorithm().id().nid();
+        let secret_key_size = self.secret_key_size();
+        let public_key_size = self.public_key_size();
+
+        if ikm.len() == self.seed_length_for_derive() {
+            Ok(unsafe { kem_derive(nid, ikm, secret_key_size, public_key_size) }?)
+        } else {
+            let ikm = self.kdf.expand(ikm, &[], self.seed_length_for_derive())?;
+            Ok(unsafe { kem_derive(nid, &ikm, secret_key_size, public_key_size) }?)
+        }
+    }
+
+    fn public_key_validate(&self, _key: &HpkePublicKey) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn seed_length_for_derive(&self) -> usize {
+        64
+    }
+
+    fn generate(&self) -> Result<(HpkeSecretKey, HpkePublicKey), Self::Error> {
+        let mut out = vec![0u8; self.seed_length_for_derive()];
+        aws_lc_rs::rand::fill(&mut out)?;
+
+        self.generate_deterministic(&out)
+    }
+
+    fn public_key_size(&self) -> usize {
+        match self.ml_kem {
+            MlKem::MlKem512 => 800,
+            MlKem::MlKem768 => 1184,
+            MlKem::MlKem1024 => 1568,
+        }
+    }
+
+    fn secret_key_size(&self) -> usize {
+        match self.ml_kem {
+            MlKem::MlKem512 => 1632,
+            MlKem::MlKem768 => 2400,
+            MlKem::MlKem1024 => 3168,
+        }
+    }
+
+    fn enc_size(&self) -> usize {
+        match self.ml_kem {
+            MlKem::MlKem512 => 768,
+            MlKem::MlKem768 => 1088,
+            MlKem::MlKem1024 => 1568,
+        }
+    }
+}
+
+unsafe fn kem_derive(
+    nid: i32,
+    ikm: &[u8],
+    mut secret_key_len: usize,
+    mut public_key_len: usize,
+) -> Result<(HpkeSecretKey, HpkePublicKey), Unspecified> {
+    let ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, null_mut());
+
+    if 1 != EVP_PKEY_CTX_kem_set_params(ctx, nid) || 1 != EVP_PKEY_keygen_init(ctx) {
+        EVP_PKEY_CTX_free(ctx);
+        return Err(Unspecified);
+    }
+
+    let mut pkey: *mut EVP_PKEY = null_mut();
+    let mut ikm_len = ikm.len();
+
+    if 1 != EVP_PKEY_keygen_deterministic(ctx, &mut pkey, ikm.as_ptr(), &mut ikm_len) {
+        EVP_PKEY_CTX_free(ctx);
+        return Err(Unspecified);
+    }
+
+    let mut secret_key = vec![0u8; secret_key_len];
+    let mut public_key = vec![0u8; public_key_len];
+
+    let err = 1 != EVP_PKEY_get_raw_private_key(pkey, secret_key.as_mut_ptr(), &mut secret_key_len)
+        || 1 != EVP_PKEY_get_raw_public_key(pkey, public_key.as_mut_ptr(), &mut public_key_len);
+
+    EVP_PKEY_CTX_free(ctx);
+    EVP_PKEY_free(pkey);
```

Comment:
Should the key be freed before the context? The key was created using the context and I don't know if the context is used when freeing the key.

Reply:
I did run it with sanitizer - wouldn't it catch problems with that?

So the doc for EVP_PKEY_keygen says "it sets |*out_pkey| to a newly-allocated |EVP_PKEY|" which I understood as "allocates independent key". I also found a couple of examples that free context and key [here](https://github.com/aws/aws-lc/blob/5084e19727f5c922a25361bb639b73d266cd8948/util/benchmark/ec/p256_awslc_ossl/src/benchmark_ecdh.c#L171) and [here](https://github.com/aws/aws-lc/blob/5084e19727f5c922a25361bb639b73d266cd8948/crypto/fipsmodule/kem/README.md?plain=1#L105)

I can also ask the guy who implemented this for us - do you think I should?

Reply:
The examples you linked both seem to free the key before the context, which is the opposite order of what is done in this PR. I don't know if sanitizer would catch such an issue. FWIW I had a quick look at the C code and I didn't see any context stored in the key or used when freeing it, so maybe the destruction order does not matter?

Asking the implementer would remove any doubt, but is it worth it? I would just follow the order in the examples you linked to be safe, but I am not opposed to merging this as-is if you think that is not worth the trouble -- your extensive testing didn't reveal any issue.